sim-lib/src/sim_node.rs

```diff
@@ -1035,3 +1038,742 @@ impl UtxoLookup for UtxoValidator {
+        assert!(channel_state.add_outgoing_htlc(hash_1, htlc_1).is_ok());
+        assert_channel_balances!(
+            channel_state,
+            local_balance - htlc_1.amount_msat,
+            1,
+            htlc_1.amount_msat
+        );
+
+        // Try to add a htlc with the same payment hash and assert that we fail because we enforce one htlc per hash
+        // at present.
+        assert!(matches!(
+            channel_state.add_outgoing_htlc(hash_1, htlc_1),
+            Err(ForwardingError::PaymentHashExists(_))
+        ));
+
+        // Add a second, distinct htlc to our in-flight state.
+        let hash_2 = PaymentHash { 0: [2; 32] };
+        let htlc_2 = Htlc {
+            amount_msat: 1000,
+            cltv_expiry: 40,
+        };
+
+        assert!(channel_state.check_outgoing_addition(&htlc_2).is_ok());
+        assert!(channel_state.add_outgoing_htlc(hash_2, htlc_2).is_ok());
+        assert_channel_balances!(
+            channel_state,
+            local_balance - htlc_1.amount_msat - htlc_2.amount_msat,
+            2,
+            htlc_1.amount_msat + htlc_2.amount_msat
+        );
+
+        // Remove our second htlc with a failure so that our in-flight drops and we return the balance.
+        assert!(channel_state.remove_outgoing_htlc(&hash_2).is_ok());
+        channel_state.settle_outgoing_htlc(htlc_2.amount_msat, false);
+        assert_channel_balances!(
+            channel_state,
+            local_balance - htlc_1.amount_msat,
+            1,
+            htlc_1.amount_msat
+        );
+
+        // Try to remove the same htlc and assert that we fail because the htlc can't be found.
+        assert!(matches!(
+            channel_state.remove_outgoing_htlc(&hash_2),
+            Err(ForwardingError::PaymentHashNotFound(_))
+        ));
+
+        // Finally, remove our original htlc with success and assert that our local balance is accordingly updated.
+        assert!(channel_state.remove_outgoing_htlc(&hash_1).is_ok());
+        channel_state.settle_outgoing_htlc(htlc_1.amount_msat, true);
+        assert_channel_balances!(channel_state, local_balance - htlc_1.amount_msat, 0, 0);
+    }
+
+    /// Tests policy checks applied when forwarding a htlc over a channel.
+    #[test]
+    fn test_htlc_forward() {
+        let local_balance = 140_000;
+        let channel_state = ChannelState::new(create_test_policy(local_balance / 2), local_balance);
+
+        // CLTV delta insufficient (one less than required).
+        assert!(matches!(
+            channel_state.check_htlc_forward(channel_state.policy.cltv_expiry_delta - 1, 0, 0),
+            Err(ForwardingError::InsufficientCltvDelta(_, _))
+        ));
+
+        // Test insufficient fee.
+        let htlc_amount = 1000;
+        let htlc_fee = (channel_state.policy.base_fee as f64
+            + ((channel_state.policy.fee_rate_prop as f64 * htlc_amount as f64) / 1000000.0))
+            as u64;
+
```

Comment:
Are there any concerns about the precision of `f64` (especially for the `fee_rate`) or is it too early to have such worries? On my device for example, a 4.7 `f64` value is sometimes represented as 4.700000002. If you're open to exploring other options, [rust_decimal](https://crates.io/crates/rust_decimal) comes highly recommended (and used) in the community and I'd be happy to work on replacing the built-in decimal type with something more precise for financial calculations